snorkel/contrib/learning/relstm.py

```diff
@@ -0,0 +1,231 @@
+
+    def _mark_sentence(self, s, mids):
+        """Insert markers around relation arguments in word sequence
+        Example: Then Barack married Michelle.  
+             ->  Then [[0 Barack 0]] married [[1 Michelle 1]].
+        """
+        marks = sorted([y for m in mids for y in self._mark(*m)], reverse=True)
+        x = list(s)
+        for k, v in marks:
+            x.insert(k, v)
+        return x
+
+    def _preprocess_data(self, candidates, extend=False):
+        """Convert candidate sentences to lookup sequences"""
+        sentences = []
+        for c in candidates:
+            # Get arguments and lemma sequence
+            args = [
+                (c[0].get_word_start(), c[0].get_word_end(), 1),
+                (c[1].get_word_start(), c[1].get_word_end(), 2)
+            ]
+            s = self._mark_sentence(
+                [w.lower() for w in c.get_parent().lemmas], args
+            )
+            # Either extend word table or retrieve from it
+            retriever = self.word_dict.get if extend else self.word_dict.lookup
+            sentences.append(np.array([retriever(w) for w in s]))
+        return sentences
+
+    def _make_tensor(self, x):
+        """Construct input tensor with padding"""
+        batch_size = len(x)
+        tx = np.zeros((self.mx_len, batch_size), dtype=np.int32)
+        tlen = np.zeros(batch_size, dtype=np.int32)
+        # Pad or trim each x
+        # TODO: fix for arguments outside max length
+        for k, u in enumerate(x):
+            lu = min(len(u), self.mx_len)
+            tx[0:lu, k] = u[0:lu]
+            tx[lu:, k] = 0
+            tlen[k] = lu
+        return tx, tlen
+
+    def _build(self):
+        """Get feed forward step, loss function, and optimizer for LSTM"""
+        # Define input layers
+        self.sentences = tf.placeholder(tf.int32, [None, None])
+        self.sentence_length = tf.placeholder(tf.int32, [None])
+        self.y = tf.placeholder(tf.float32, [None])
+        # Embedding layer
+        embedding = tf.Variable(
+            tf.random_normal((self.n_v, self.dim), mean=0, stddev=1.0),
+            name='embedding'
+        )
+        inputs = tf.nn.embedding_lookup(embedding, self.sentences)
+        # Get RNN graph
+        batch_size = tf.shape(self.sentences)[1]
+        with tf.variable_scope("LSTM") as scope:
+            # LSTM architecture
+            cell = tf.nn.rnn_cell.BasicLSTMCell(self.dim, state_is_tuple=True)
+            # Set RNN
+            initial_state = cell.zero_state(batch_size, tf.float32)
+            rnn_outputs, _ = tf.nn.dynamic_rnn(
+                cell, inputs, initial_state=initial_state, time_major=True
+            )
+            # Get LSTM variables
+            self.save_dict = {
+                x.name: x for x in tf.get_collection(
+                tf.GraphKeys.GLOBAL_VARIABLES, scope=scope.name)
+            }
+        # Take mean across sentences
```

Comment:
Yeah, simplest sentence embedding technique. Can improve down the road, but easy enough for now.